### ISSUE TITLE
[wiki] Summarize Implemented Methods

### ISSUE THREAD

**Issue by Jingkang50** [Jingkang50/OpenOOD#62]
We plan to write a short summary for every implemented method in our wiki space.
Please follow the template below and write the draft here.

**Paper Title**

[![paper](https://img.shields.io/badge/Paper-arxiv-b31b1b)](paper_link) &nbsp; [![code](https://img.shields.io/badge/Code-github-52b788)](original_github_link)

**Method Description**

> - **Overview:** A one-sentence summary
> - **Model Architecture:** special design for network architectures
> - **Training:** special design for training pipeline
> - **Inference:** special design for inference pipeline
> - **Comments:** say something more if any

**Implementation** (List all the related python files)

> - `xxx.py:` the function of this python file.

**Script**
```bash
# how to run the code
```

**Comment by Zzitang** [comment]
**CutPaste: Self-Supervised Learning for Anomaly Detection and Localization**

[![paper](https://img.shields.io/badge/Paper-arxiv-b31b1b)](http://arxiv.org/abs/2104.04015) &nbsp; [![code](https://img.shields.io/badge/Code-github-52b788)](original_github_link)

**Method Description**

> - **Overview:**  CutPaste is a data augmentation strategy that constructs anomaly patterns from normal ones. A one-class classifier is built to distinguish anomalous data from normal ones in this method. And Gaussian density estimator is used to compute the anomaly score.
> - **Training:** Use cutpaste preprocessor to generate anomalous images by cutting an image patch and pastes at a random location.
> - **Inference:** Use Gaussian density estimator to compute the anomaly score.

**Implementation** (List all the related python files)

> - `train_ad_pipeline.py:` train pipeline.
> - `test_ad_pipeline.py:` test pipeline.
> - `cutpaste_trainer.py:` trainer for cutpaste to train an one class classifier.
> - `cutpaste_evaluator.py:` evaluator.
> - `cutpaste_preprocessor.py:` preprocessor to perform CutPaste data augmentation.
> - `cutpaste_postprocessor.py:` postprocessor to calculate anomaly score by Gaussian density estimator.
> - `cutpaste_recorder.py:` recorder.

**Script**
```bash
> - `cutpaste_train` sh scripts/_get_started/1_mnist_test.sh
> - `cutpaste_test` sh scripts/a_anomaly/2_cutpaste_train.sh
# how to run the code
```

**Comment by Zzitang** [comment]
**Detecting Out-of-Distribution Examples with In-distribution Examples and Gram Matrices**

[![paper](https://img.shields.io/badge/Paper-arxiv-b31b1b)](http://arxiv.org/abs/1912.12510) &nbsp; [![code](https://img.shields.io/badge/Code-github-52b788)](original_github_link)

**Method Description**

> - **Overview:**  Gram detects OOD examples by identifying anomalies in the gram matrices.
> - **Inference:** Use Gaussian density estimator to compute the anomaly score.

**Implementation** (List all the related python files)

> - `test_ood_pipeline.py:` test pipeline.
> - `ood_evaluator.py:` evaluator.
> - `gram_postprocessor.py:` postprocessor to calculate gram metrics deviations from in-distribution data .

**Script**
```bash
# how to run the code
sh scripts/c_ood/7_cifar_test_ood_gram.sh
```

**Comment by JediWarriorZou** [comment]
DeepSVDD 
**Method Description**
> - **Pretrain:**
During this stage, we pretrain a deep convolutional autoencoders (dcae) for anomaly detection.
> - **Train:**
During this stage, firstly we load the pretrained dcae into the net, then we proceed to train the model.
The output includes net and two hyperparameters C and R which represent center and radius of the hypersphere.
> - **Test:**
During this stage, we evaluate our method with metric auroc.

**OpenOOD Implementation**
> - `train_ad_pipeline.py:` Pretrain the dcae
> - `train_dsvdd_pipeline.py:`Train our model and finally test our model
> - `dsvdd_net.py:` Define dcae net and dsvdd net
> - `dsvdd_trainer.py:` Trainer of dcae and dsvdd
> - `dsvdd_evaluator.py:` Evaluator of dcae and dsvdd

**Script**
> - `pretrain dcase:`sh scripts/a_anomaly/0_dsvdd_pretrain.sh
> - `train dsvdd:`sh scripts/a_anomaly/0_dsvdd_train.sh

**Result**
> - **Note:**
In the original code of dsvdd, the train dataset is normalized with special means and stds. For example, when normal dataset is cifar10-3, normalization dict:  [-31.7975, -31.7975, -31.7975], [42.8907, 42.8907, 42.8907]. Furthermore, a global_contrast_normalization method is used in the transform. So the ideal result is displayed below.


| Normal class   | 3     | 3     |
|----------------|-------|-------|
| Method         | DCAE  | DSVDD |
| Expected AUROC | 58.40 | 59.10 |
| AUROC          | 63.43 | 60.44 |

**Comment by JediWarriorZou** [comment]
**Method Description**:
> - **train:**
 During the training stage, we introduce two vgg networks, one of which called source network is pretrained. For each training epoch, when id data is input, the differences in special layers between clone nwtwork and source network are obtained and loss is computed. By SGD, the clone network is optimised.
> - **test:**
During the testing stage, we evaluate the method of anomaly detection by roc_auc.

**Keypoints**:
> - `train_ad_pipeline.py:` training stage
> - `ad_test_pipeline.py:` testing stage
> - `kdad_trainer.py:` trainer
> - `kdad_evaluator.py:` evaluator
> - `vggnet.py:` source and clone network
> - `kdad_recorder.py:` recorder
> - `kdad_losses.py:` define loss function

**Script**
> - `kdad_train:` sh scripts/a_anomaly/1_kdad_train.sh
> - `kdad_detection_test:` sh scripts/a_anomaly/1_kdad_test_det.sh

**Result**
| Normal class   | 3     | 
|----------------|-------|
| Expected AUROC | 77.02 | 
| AUROC          | 86.08 |

**Comment by JediWarriorZou** [comment]
Confidence Branch
**Method Description**
> - **train:**
During the training stage, the net output includes scores and confidence. The loss is composed of classification loss and confidence loss.
> - **train:**
During the test stage, we evaluate the model for OOD detection.

**Keypoints**
> - **conf_widernet.py:** wideresnet that can output confidence
> - **conf_esti_trainer.py:** trainer
> - **conf_esti_recorder.py:** recorder
> - **conf_esti_evaluator.py:** evaluator
> - **train_pipeline.py:** train pipeline for classification
> - **test_ad_pipeline.py:** test pipeline for OOD detection

**Script**
> -  "Train:" sh scripts/e_conf/train_conf_esti.sh
> -  "Test:" sh scripts/e_conf/test_conf_esti.sh

**Result**
| dataset       | svhn   |
|---------------|--------|
| Eval Acc      | 0.966  |
| fpr95         | 0.2516 |
| auroc         | 0.9541 |
| aupr_in       | 0.9824 |
| aupr_out      | 0.8757 |
| detection_err | 0.1103 |

**Comment by JediWarriorZou** [comment]
VOS
**Method Description**
> - **train:**
During the training stage, virtual outliers are generated based on penultimate layer of the network and model loss comprises classification loss and uncertainty loss.
> - **test:**
During the testing stage, we evaluate the model for OOD detection.
**Keypoints**
> - **wrn.py:** wideresnet is used
> - **vos_trainer.py:** trainer
> - **base_recorder.py:** recorder
> - **vos_evaluator.py:** evaluator
> - **train_pipeline.py:** train pipeline for classification
> - **test_ad_pipeline.py:** test pipeline for OOD detection

**Script**
> -  **Train:** sh scripts/c_ood/13_cifar_train_vos.sh
> -  **Test:** sh scripts/c_ood/13_cifar_test_vos.sh

**Result**
| Metric        | Value/% |
|---------------|---------|
| VAL_ACC       | 94.55   |
| fpr95         | 47.71   |
| AUROC         | 93.73   |
| AUPR_IN       | 91.45   |
| AUPR_OUT      | 96.40   |
| DETECTION_ERR | 11.56   |

**Comment by OmegaDING** [comment]
**MOS: Towards Scaling Out-of-distribution Detection for Large Semantic Space**

[![paper](https://img.shields.io/badge/Paper-arxiv-b31b1b)](https://arxiv.org/abs/2105.01879) &nbsp; [![code](https://img.shields.io/badge/Code-github-52b788)](https://github.com/deeplearning-wisc/large_scale_ood)

**Method Description**

> - **Overview:** Obtain the ood scores of different categories, and then detect them separately
> - **Model Architecture:** Bit model with little difference
> - **Training:** need to calculate group softmax loss 
> - **Inference:** need to load group slices while inference

**Implementation** (List all the related python files)

> - `mos_evaluator.py:` evaluator
> - `mos_network.py:` network 
> - `mos_evaluator.py:` evaluator
> - `test_mos_pipeline.py:` test pieline
> - `train_pipeline.py:` train pipeline
> - `mos_trainer.py:` trainer


**Script**
```bash
# train
sh scripts/c_ood/mos.sh
# test
sh scripts/c_ood/test_mos.sh

```
